sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StatePartitionAllColumnFamiliesWriterSuite.scala

```diff
@@ -468,15 +593,373 @@ class StatePartitionAllColumnFamiliesWriterSuite extends StateDataSourceTestBase
+          StartStream(checkpointLocation = sourceDir.getAbsolutePath,
+            trigger = Trigger.ProcessingTime("1 second"),
+            triggerClock = clock),
+          AddData(inputData, "a"),
+          AdvanceManualClock(1 * 1000),
+          CheckNewAnswer(("a", "1")),
+          StopStream
+        )
+
+        // Step 2: Create target checkpoint with dummy data
+        val clock2 = new StreamManualClock
+        testStream(result, OutputMode.Update())(
+          StartStream(checkpointLocation = targetDir.getAbsolutePath,
+            trigger = Trigger.ProcessingTime("1 second"),
+            triggerClock = clock2),
+          AddData(inputData, "x"),
+          AdvanceManualClock(1 * 1000),
+          CheckNewAnswer(("a", "1"), ("x", "1")),
+          StopStream
+        )
+
+        // Step 3: Build timer column family configs and perform round-trip test
+        val (columnFamilyToSchemaMap, columnFamilyToSelectExprs, columnFamilyToStateSourceOptions) =
+          getTimerStateConfigsForCountState(TimeMode.ProcessingTime())
+
+        performRoundTripTest(
+          sourceDir.getAbsolutePath,
+          targetDir.getAbsolutePath,
+          columnFamilyToSchemaMap,
+          columnFamilyToSelectExprs = columnFamilyToSelectExprs,
+          columnFamilyToStateSourceOptions = columnFamilyToStateSourceOptions
+        )
+      }
+    }
+  }
+
+  /**
+   * Helper method to test round-trip for transformWithState with list state and TTL.
+   */
+  private def testListStateTTLRoundTrip(): Unit = {
+    withTempDir { sourceDir =>
+      withTempDir { targetDir =>
+        val clock = new StreamManualClock
+        val inputData = MemoryStream[String]
+        val result = inputData.toDS()
+          .groupByKey(x => x)
+          .transformWithState(new ListStateTTLProcessor(),
+            TimeMode.ProcessingTime(),
+            OutputMode.Update())
+
+        // Step 1: Create source checkpoint
+        testStream(result, OutputMode.Update())(
+          StartStream(checkpointLocation = sourceDir.getAbsolutePath,
+            trigger = Trigger.ProcessingTime("1 second"),
+            triggerClock = clock),
+          AddData(inputData, "a", "b", "a"),
+          AdvanceManualClock(1 * 1000),
+          CheckNewAnswer(("a", "2"), ("b", "1")),
+          StopStream
+        )
+
+        // Step 2: Create target checkpoint with dummy data
+        val clock2 = new StreamManualClock
+        testStream(result, OutputMode.Update())(
+          StartStream(checkpointLocation = targetDir.getAbsolutePath,
+            trigger = Trigger.ProcessingTime("1 second"),
+            triggerClock = clock2),
+          AddData(inputData, "x"),
+          AdvanceManualClock(1 * 1000),
+          CheckNewAnswer(("a", "2"), ("b", "1"), ("x", "1")),
+          StopStream
+        )
+
+        // Step 3: Define schemas for list state with TTL column families
+        val groupByKeySchema = StructType(Array(
```

Comment:
ditto, lets try to see as much things we can move to a common test util. So we can reuse them easily for new test suites